proposals/NNNN-codable-error-printing.md

```diff
@@ -0,0 +1,173 @@
+
+## Future directions
+
+### Better error generation from Foundation encoders/decoders
+
+The debug descriptions generated in Foundation sometimes contain the same information as the new debug descriptions from this proposal. A future change to the standard JSON and Plist encoders and decoders could provide more compact debug descriptions once they can be sure they have the new standard library descriptions available. They could also use a more compact description when rendering the description of a `CodingKey`. Using part of the example from above:
+
+```
+Debug description: No value associated with key CodingKeys(stringValue: "population", intValue: nil) ("population").
+```
+
+The `CodingKeys(stringValue: "population", intValue: nil) ("population")` part is coming from the default `description` of `CodingKey`, plus an extra parenthesized string value at the end for good measure. The Foundation (de|en)coders could construct a more compact description that does not repeat the key, just like we do within this proposal in the context of printing a coding path.
+
+### Print context of surrounding lines in source data
+
+When a decoding error occurs, in addition to printing the path, the error message could include some surrounding lines from the source data. This was explored in this proposal's antecedent, [UsefulDecode](https://github.com/ZevEisenberg/UsefulDecode). But that requires passing more context data from the decoder and changing the public interface of `DecodingError` to carry more data. This option is probably best left as something to think about as [we design `Codable`'s successor](https://forums.swift.org/t/the-future-of-serialization-deserialization-apis/78585). But just to give an example of the _kind_ of context that could be provided (please do not read anything into the specifics of the syntax; this is a sketch, not a proposal):
+
+```
+Value not found: expected 'name' (String) at [0]/address/city/birds/[1]/name, got:
+{
+  "feathers" : "some",
+  "name" : null
+}
+```
+
+## Alternatives considered
+
+The original version of this proposal suggested conforming `EncodingError` and `DecodingError` to `CustomStringConvertible`, not `CustomDebugStringConvertible`. The change to the debug-flavored protocol emphasizes that the new descriptions aren't intended to be used outside debugging contexts. This is in keeping with the precedent set by [SE-0445](0445-string-index-printing.md).
+
+The original version also proposed changing `CodingKey.description` to return the bare string or int value, but changing the exsting implementation of an existing public method was deemed too potentially dangerous.
```

Comment:
Can you specify what you mean here by "dangerous"?

Reply:
It was in reference to https://github.com/swiftlang/swift/pull/80941#discussion_r2064277369. I'll update the proposal, plus I'll link to that thread if appropriate.

Reply:
So, if I understand the point correctly, "dangerous" is referring concretely to `Foundation.JSONDecoder` having a known dependency on non-guaranteed implementation details of `CodingKey.description`?

Reply:
yep! I'm thinking "dangerous" is not the right word, and @kperryua's original message or your summary is closer to what we should put here in the doc.